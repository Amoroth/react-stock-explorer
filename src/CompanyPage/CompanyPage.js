import React, { Component } from 'react'
import PropTypes from 'prop-types'

import PageTitle from '../shared/PageTitle'
import styles from './CompanyPage.module.css'

class CompanyPage extends Component {
  state = {
    company: {
      companyName: '',
      symbol: '',
      exchange: '',
      industry: '',
      website: '',
      description: '',
      CEO: '',
    },
    logo: '',
    financials: [
      {
        reportDate: '',
        grossProfit: 0,
        totalRevenue: 0,
        netIncome: 0,
        researchAndDevelopment: 0,
        totalAssets: 0,
        totalDebt: 0,
      },
    ],
    finTime: 0,
  }

  componentDidMount() {
    const { location } = this.props
    const companyQuery = new URLSearchParams(location.search.slice(1)).get(
      'cmp',
    )
    fetch(
      `https://api.iextrading.com/1.0/stock/${companyQuery}/batch?types=company,logo,financials`,
    )
      .then((res) => res.json())
      .then((json) => {
        this.setState({
          company: json.company,
          logo: json.logo.url,
          financials: json.financials.financials,
        })
      })
  }

  onSelectFintime = (index) => {
    this.setState({ finTime: index })
  }

  render() {
    const { financials, finTime, logo, company } = this.state
    const { favorites, onFavorite, currency, currencyFormat } = this.props

    let currentReport = {}
    if (financials) {
      currentReport = Object.assign({}, financials[finTime])
      delete currentReport.reportDate
    }

    let financialsElement = null
    if (financials) {
      financialsElement = (
        <React.Fragment>
          <hr />
          <div>
            <div className={styles['fintime-container']}>
              {financials.map((val, ind) => (
                <button
                  className={`${styles['fintime-button']} ${
                    finTime === ind ? styles['fintime-button-active'] : null
                  }`}
                  type="button"
                  onClick={() => this.onSelectFintime(ind)}
                  key={val.reportDate}
                >
                  {val.reportDate}
                </button>
              ))}
            </div>
            <div className={styles['fin-info']}>
              {Object.keys(currentReport).map((val) => {
                const tempName = val.replace(/([A-Z])/g, ' $1')
                const labelName = tempName.charAt(0).toUpperCase() + tempName.slice(1)
                return (
                  <React.Fragment key={val}>
                    <span>{labelName}</span>
                    <span>{currentReport[val] ? `${currencyFormat(currentReport[val])} ${currency}` : '-'}</span>
                  </React.Fragment>
                )
              })}
            </div>
          </div>
        </React.Fragment>
      )
    }

    return (
      <div className={styles.container}>
        <PageTitle
          logo={logo}
          name={company.companyName}
          exchange={company.exchange}
          symbol={company.symbol}
          favorite={favorites.includes(company.symbol)}
          onFav={onFavorite}
        />
        <hr />
        <div>
          <div className={styles['company-info-row']}>
            <span>
              {`CEO: ${company.CEO}`}
            </span>
            <span>
              {`Industry: ${company.industry}`}
            </span>
            <span>
              <a href={company.website}>Website</a>
            </span>
          </div>
          <p>{company.description || 'No Description.'}</p>
        </div>
        {financialsElement}
      </div>
    )
  }
}

CompanyPage.propTypes = {
  location: PropTypes.objectOf(PropTypes.string).isRequired,
  favorites: PropTypes.arrayOf(PropTypes.string).isRequired,
  onFavorite: PropTypes.func.isRequired,
  currency: PropTypes.string.isRequired,
  currencyFormat: PropTypes.func.isRequired,
}

export default CompanyPage
