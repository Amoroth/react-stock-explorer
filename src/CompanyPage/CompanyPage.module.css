.container {
  margin: auto;
  width: 60%;
  max-width: 800px;
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 2px 4px 10px rgba(0,0,0,.025);
  padding: 16px;
}

.company-info-row {
  display: flex;
  justify-content: space-between;
}

.container a {
  color: #344955;
  transition: color .2s;
}

.container a:hover {
  color: #75b8e2;
}

@media screen and (max-width: 620px) {
  .container {
    width: 97%;
  }

  .company-info-row {
    flex-direction: column;
  }

  .company-info-row span {
    margin: 10px 0;
  }
}