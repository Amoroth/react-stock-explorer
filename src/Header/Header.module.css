.container {
  align-items: center;
  background-color: #344955;
  box-shadow: 2px 4px 10px rgba(0,0,0,.2);
  display: flex;
  height: 62px;
  justify-content: space-between;
  position: fixed;
  top: 0;
  width: 100%;
}

.brand {
  color: #fff;
  font-size: 22px;
  margin: 0 15px;
}

.nav-links {
  height: 100%;
  margin: 0 15px;
}

.nav-links a {
  color: #fff;
  display: inline-block;
  height: 100%;
  line-height: 60px;
  margin-left: 15px;
  position: relative;
  text-decoration: none;
}

.nav-links a:hover {
  color: #ddd;
}

.nav-link-active {
  border-bottom: 5px solid #fff;
}
.nav-link-active:hover {
  border-bottom: 5px solid #ddd;
}

@media screen and (max-width: 620px) {
  .nav-links {
    visibility: hidden;
  }
}