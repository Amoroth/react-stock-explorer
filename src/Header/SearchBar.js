import React, { Component } from 'react'
import { createFilter } from 'react-search-input'
import { Link } from 'react-router-dom'

import Overlay from '../shared/Overlay'
import styles from './Header.module.css'

class SearchBar extends Component {
  constructor(props) {
    super(props)
    this.state = {
      input: '',
      symbols: [],
      searchResults: false,
      searchBoxInput: false,
    }
    this.inputRef = React.createRef();
  }

  componentDidMount() {
    fetch('https://api.iextrading.com/1.0/ref-data/symbols?filter=symbol,name').then((res) => {
      return res.json()
    }).then((json) => {
      this.setState({ symbols: json, searchBoxInput: window.innerWidth > 800 })
    })
    this.setState({ symbols: [
      {name: 'Alphabet Inc.', symbol: 'googl'},
      {name: 'Microsoft Corporation', symbol: 'msft'},
      {name: 'International Buissnes Machines Corporation', symbol: 'ibm'},
      {name: 'Twitter Inc.', symbol: 'twtr'}
    ],
    searchBoxInput: window.innerWidth > 800 })
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.input !== prevState.input) {
      return true
    }
    return false
  }

  handleChange = (e) => {
    this.setState({ input: e.target.value })
  }

  onClickFill = (term) => {
    this.setState({ input: term })
  }

  showSearchResults = () => {
    this.setState({ searchResults: true })
  }

  hideSearchResults = () => {
    setTimeout(() => this.setState({ searchResults: false, searchBoxInput: window.innerWidth > 800 }), 100)
  }

  showSearchBoxInput = () => {
    this.setState({ searchBoxInput: true }, () => {
      this.inputRef.current.focus()
    })
  }

  render() {
    const filteredCmps = this.state.symbols.filter(
      createFilter(this.state.input, ['name', 'symbol'])
    )

    return (
      <div style={{display: 'inline'}}>
        {this.state.searchBoxInput ? 
        <input
          className={styles['search-box']}
          onChange={this.handleChange}
          value={this.state.input}
          onFocus={this.showSearchResults}
          onBlur={this.hideSearchResults}
          ref={this.inputRef}
        /> : 
        <button onClick={this.showSearchBoxInput} className={styles['search-box-icon']}>
          <i className="material-icons">search</i>
        </button>}

        <div className={styles['search-box-results']}>
          {this.state.searchResults ? filteredCmps.slice(0, 4).map((cmp) => {
            return (
              <Link
                className={styles['search-box-result']}
                key={cmp.symbol}
                onClick={() => this.onClickFill(cmp.name)}
                to={`stock?cmp=${cmp.symbol}`}>
                {cmp.name}
                <span>{cmp.symbol}</span>
              </Link>
            )}) : null}
        </div>
        {this.state.searchResults && window.innerWidth < 800 ? <Overlay /> : null}
      </div>
    )
  }
}

export default SearchBar