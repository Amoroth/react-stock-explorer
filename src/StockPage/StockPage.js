import React, { Component } from 'react'
import Draggable from 'react-draggable'
import { Link } from 'react-router-dom'

import PageTitle from '../shared/PageTitle'
import StockChart from './StockChart'
import Spinner from '../shared/Spinner'
import styles from './StockPage.module.css'

class StockPage extends Component {
  state = {
    book: {
      symbol: '',
      companyName: '',
      primaryExchange: '',
      sector: '',
      open: 0,
      openTime: 0,
      close: 0,
      closeTime: 0,
      high: 0,
      low: 0,
      previousClose: 0,
      change: 0,
      week52High: 0,
      week52Low: 0
    },
    chart: [],
    chartTime: 0,
    relevant: [],
    logo: ''
  }

  componentDidMount() {
    const cmp = new URLSearchParams(this.props.location.search.slice(1)).get('cmp')
    this.updateBook(cmp)
    this.updateChart(cmp, '1m')
  }

  componentDidUpdate() {
    const cmp = new URLSearchParams(this.props.location.search.slice(1)).get('cmp')
    if (cmp.toLowerCase() !== this.state.book.symbol.toLowerCase()) {
      this.updateBook(cmp)
      this.updateChart(cmp, '1m')
    }
  }

  updateBook = (cmp) => {
    fetch(`https://api.iextrading.com/1.0/stock/${cmp}/batch?types=quote,relevant,logo`)
      .then((res) => {
        return res.json()
      }).then((json) => {
        fetch(`https://api.iextrading.com/1.0/stock/market/batch?types=quote&symbols=${json.relevant.symbols.join(',')}&filter=companyName,symbol,changePercent,close`)
          .then((response) => {
            return response.json()
          }).then((endJson) => {
            let relevantCompanies = []
            for (let key in endJson) {
              let tempCompany = {...endJson[key].quote}
              tempCompany.symbol = key
              relevantCompanies.push(tempCompany)
            }
            this.setState({
              book: json.quote,
              relevant: relevantCompanies,
              logo: json.logo.url
            })
          })
      })
  }

  updateChart = (symbol, time) => {
    fetch(`https://api.iextrading.com/1.0/stock/${symbol}/chart/${time}?filter=date,close,label`).then((res) => {
      return res.json()
    }).then((json) => {
      let timeNum = ['1m', '3m', '6m', '1y', '2y'].indexOf(time)
      this.setState({
        chart: json,
        chartTime: timeNum
      })
    })
  }

  onSelectCharttime = (time) => {
    this.setState({
      chart: []
    })
    this.updateChart(this.state.book.symbol, time)
  }

  render() {
    let chartEl = (
      <div className={styles['spinner-container']}>
        <Spinner />
      </div>
    )
    if (this.state.chart.length > 0) {
      chartEl = <StockChart data={this.state.chart} />
    }

    return (
      (
        <div className={styles['container']}>
          <PageTitle
            logo={this.state.logo}
            name={this.state.book.companyName}
            exchange={this.state.book.primaryExchange}
            symbol={this.state.book.symbol}
            link
          />
          <hr />
          <div>
            <div className={styles['charttime-container']}>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 0 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('1m')}>1m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 1 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('3m')}>3m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 2 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('6m')}>6m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 3 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('1y')}>1y</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 4 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('2y')}>2y</button>
            </div>
            { chartEl }
            <div className={styles['book-info']}>
                <span>Open:</span> <span>{this.state.book.open} USD</span>
                <span>Prev. Close:</span> <span>{this.state.book.previousClose} USD</span>
                <span>Max:</span> <span>{this.state.book.high} USD</span>
                <span>Highest/52 weeks:</span> <span>{this.state.book.week52High} USD</span>
                <span>Min:</span> <span>{this.state.book.low} USD</span>
                <span>Lowest/52 weeks</span> <span>{this.state.book.week52Low} USD</span>
                <span>Market Cap</span> <span>{(this.state.book.marketCap / 1000000000).toFixed(2)} mld USD</span>
                <span>P/E Ratio</span> <span>{this.state.book.peRatio} USD</span>
            </div>
          </div>
          <hr />
          <div>
            {this.state.book.sector ? <p>Sector: {this.state.book.sector}</p> : <p>Relevant:</p>}
            <div className={styles['relevant']}>
              <Draggable
                axis="x"
                defaultPosition={{x: 0, y: 0}}
                bounds={{left: -1010, right: 0}}
              >
                <div className={styles['relevant-inner']}>
                  {this.state.relevant.map((val) => {
                    let shortCompanyName = val.companyName
                    let priceChange = `${val.changePercent > 0 ? '+' : ''}${(val.changePercent * 100).toFixed(2)}%`

                    if (shortCompanyName.length > 14)
                      shortCompanyName = shortCompanyName.slice(0, 12) + '...'
                    return (
                      <div className={styles['relavent-item']} key={val.symbol}>
                        <div className={`${styles['relavent-item-row']} ${styles['relavent-item-title']}`}>
                          <h6 title={ shortCompanyName[shortCompanyName.length - 2] === '.' ? val.companyName : null }>{shortCompanyName}</h6>
                          <span>{val.symbol}</span>
                        </div>
                        <div className={`${styles['relavent-item-row']} ${styles['relavent-item-info']}`}>
                          <span>{val.close}</span>
                          <span style={{color: val.changePercent > 0 ? 'green' : 'red'}}>
                            { priceChange }
                          </span>
                          <Link
                            to={`/stock?cmp=${val.symbol}`}
                            onDragStart={(e) => e.preventDefault()}
                            onClick={() => window.scrollTo(0, 0)}
                          >More...</Link>
                        </div>
                      </div>
                    )
                  })}
                </div>
              </Draggable>
            </div>
          </div>
          <hr />
          <div>
            <p>Informacje Finansowe (osobne okno?)</p>
            <p>O firmie (osobne okno?)</p>
          </div>
        </div>
      )
    )
  }
}

export default StockPage