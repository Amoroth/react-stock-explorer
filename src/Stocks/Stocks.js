import React from 'react'

import Panel from './Panel'
import Spinner from '../shared/Spinner'
import styles from './Stocks.module.css'

class Stocks extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      stocks: [
        {
          companyName: 'Placeholder',
          symbol: 'none',
          close: 1098.77,
          change: 0.0467,
          favorite: false
        }
      ]
    }
  }

  componentDidMount() {
    fetch('https://api.iextrading.com/1.0/stock/market/list/infocus').then((res) => {
      return res.json()
    }).then((json) => {
      this.setState({
        stocks: json
      })
    })

    this.loadFavorites()
  }

  loadFavorites = () => {
    console.log('loading favorites')

    // fetch('https://api.iextrading.com/1.0/stock/market/batch?types=previous&symbols=aapl,msft,googl')
  }

  onFavorite = (event, symbol) => {
    event.stopPropagation()

    console.log('added to favorites')

    // let newStocks = this.state.stocks.slice()
    // newStocks.forEach(element => {
    //   if (element.short === symbol)
    //     element.favorite = !element.favorite
    // })
    // this.setState({
    //   stocks: newStocks
    // }, () => {
    //   localStorage.setItem('stock-explorer_favorites', this.state.stocks.reduce((arr, val) => {
    //     if (val.favorite)
    //       arr.push(val.symbol)
    //     return arr
    //   }, []))
    // })
  }

  render() {
    const stocks = this.state.stocks.map((val) => {
      return <Panel 
        title={val.companyName} 
        short={val.symbol} 
        price={val.close} 
        change={val.changePercent} 
        // favorite={val.favorite}
        key={val.symbol}
        onFav={this.onFavorite}
      />
    })
    return (
      <div className={styles['container']}>
        {this.props.children}
        {this.state.stocks[0].symbol === 'none' ? <Spinner /> : stocks}
      </div>
    )
  }
}

export default Stocks