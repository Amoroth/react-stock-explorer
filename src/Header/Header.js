import React from 'react'
import { withRouter, NavLink } from 'react-router-dom'

import SearchBar from './SearchBar'
import styles from './Header.module.css'

const header = (props) => {
  let expandButton = (
  <button className={styles['expand-button']}>
    <i className="material-icons">menu</i>
  </button>)

  const onBackClick = () => {
    if (props.location.pathname === '/company') {
      props.history.goBack()
    } else {
      props.history.push('/')
    }
  }

  if (props.location.pathname !== '/') {
    expandButton = (
      <button className={styles['expand-button']} onClick={onBackClick}>
        <i className="material-icons">arrow_back</i>
      </button>
    )
  }

  let navlinkStyle = [styles['nav-link'], styles['desktop-only']].join(' ')

  return (
    <nav className={styles['container']}>
      <div className={styles['brand-container']}>
        { expandButton }
        <h6 className={styles['brand']}>Stock Explorer!</h6>
      </div>
      <div className={[styles['nav-links']].join(' ')}>
        <SearchBar />
        <NavLink to="/" exact activeClassName={styles['nav-link-active']} className={navlinkStyle}>Stocks</NavLink>
        <NavLink to="/error" activeClassName={styles['nav-link-active']} className={navlinkStyle}>Cryptos</NavLink>
      </div>
    </nav>
  )
}

export default withRouter(header)