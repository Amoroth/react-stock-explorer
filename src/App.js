import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom'

import Header from './Header/Header'
import Stocks from './Stocks/Stocks'
import StockPage from './StockPage/StockPage'

class App extends Component {
  render() {
    let router = (
      <Switch>
        <Route path="/" exact component={Stocks} />
        <Route path="/stock" exact component={StockPage} />
        <Route render={() => <h1>404</h1>} />
      </Switch>
    )

    return (
      <div className="App">
        <Header />
        {router}
      </div>
    );
  }
}

export default App;
