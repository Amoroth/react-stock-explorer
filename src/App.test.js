import React from 'react'
import { mount, shallow } from 'enzyme'
import * as rrd from 'react-router-dom'

import App from './App'
import Header from './Header/Header'
import Stocks from './Stocks/Stocks'
import StockPage from './StockPage/StockPage'
import CompanyPage from './CompanyPage/CompanyPage'

// eslint-disable-next-line react/prop-types
rrd.BrowserRouter = ({ children }) => <div>{children}</div>

describe('<app />', () => {
  it('renders without crashing', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={[{ pathname: '/market', key: 'testkey' }]}>
        <App />
      </rrd.MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders "/market" route', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={[{ pathname: '/market', key: 'testkey' }]}>
        <App />
      </rrd.MemoryRouter>,
    )
    expect(wrapper.find(Stocks).exists()).toBe(true)
  })

  it('redirects from "/" to "/stocks"', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={[{ pathname: '/', key: 'testkey' }]}>
        <App />
      </rrd.MemoryRouter>,
    )
    expect(wrapper.find(Stocks).exists()).toBe(true)
  })

  it('renders "/stock" route', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={[{ pathname: '/stock', key: 'testkey' }]}>
        <App />
      </rrd.MemoryRouter>,
    )
    expect(wrapper.find(StockPage).exists()).toBe(true)
  })

  it('renders "/company" route', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={[{ pathname: '/company', key: 'testkey' }]}>
        <App />
      </rrd.MemoryRouter>,
    )
    expect(wrapper.find(CompanyPage).exists()).toBe(true)
  })

  it('renders header component', () => {
    const wrapper = mount(
      <rrd.MemoryRouter initialEntries={['/']}>
        <App />
      </rrd.MemoryRouter>,
    )
    expect(wrapper.find(Header).exists()).toBe(true)
  })

  it('correctly calls onFavorite', () => {
    const wrapper = shallow(<App />, { disableLifecycleMethods: true })

    wrapper.instance().onFavorite({ stopPropagation: jest.fn() }, 'SYMB')
    expect(wrapper.state('favorites')).toContain('SYMB')

    wrapper.instance().onFavorite({ stopPropagation: jest.fn() }, 'SYMB')
    expect(wrapper.state('favorites')).toHaveLength(0)
  })

  it('correctly changes currency', () => {
    const wrapper = shallow(<App />, { disableLifecycleMethods: true })

    wrapper.instance().onCurrencyChange('EUR')
    expect(wrapper.state('currency')).toBe('EUR')
  })

  it('correctly formats number', () => {
    const wrapper = shallow(<App />, { disableLifecycleMethods: true })
    wrapper.setState({ currencyRates: { USD: 1, EUR: 2, GBP: 3 } })

    expect(wrapper.instance().currencyFormat(10)).toBe('10.00')
    expect(wrapper.instance().currencyFormat(5000000)).toBe('5.00 mln')
    expect(wrapper.instance().currencyFormat(5000000000)).toBe('5.00 mld')

    wrapper.setState({ currency: 'EUR' })
    expect(wrapper.instance().currencyFormat(5000000)).toBe('10.00 mln')
  })

  it('correctly reads from localStorage', () => {
    localStorage.setItem('stock-explorer_favorites', ['LOL', 'OMG', 'ZTM'])
    localStorage.setItem('stock-explorer_currency', 'PLN')
    const wrapper = shallow(<App />)

    expect(wrapper.state('favorites')).toContain('LOL', 'OMG', 'ZTM')
    expect(wrapper.state('currency')).toBe('PLN')
  })
})
