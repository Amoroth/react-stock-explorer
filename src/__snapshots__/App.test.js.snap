// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<app /> renders without crashing 1`] = `
<div
  className="App"
>
  <withRouter(header)
    currency="USD"
    currencyChange={[Function]}
  />
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/market"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/stock"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/company"
      render={[Function]}
    />
    <Route
      render={[Function]}
    />
  </Switch>
</div>
`;
